Logger for dialogs

ChristoferAlexander opened the pull request:
#733

ChristoferAlexander commented:
Logs look like this for Android:

Displaying alert dialog with title: Hello, Kaluga 🐟
Action OK was called on dialog with title: Hello, Kaluga 🐟
Dismissing alert dialog with title: Hello, Kaluga 🐟
Displaying alert dialog with title: Wait for 3 sec...
Dismissing alert dialog with title: Wait for 3 sec...
Displaying alert dialog with title: Select an option
Action Option 2 was called on dialog with title: Select an option
Dismissing alert dialog with title: Select an option
Displaying alert dialog with title: Hello, Kaluga 🐟
Action OK was called on dialog with title: Hello, Kaluga 🐟
Dismissing alert dialog with title: Hello, Kaluga 🐟

Daeda88 commented on alerts/src/iosMain/kotlin/AlertPresenter.kt:
```diff
@@ -143,7 +153,11 @@ actual class AlertPresenter(
-            parent.presentViewController(this, animated, completion)
+            logger.info(TAG, "Displaying alert dialog with title: ${alert.title}")
+            parent.presentViewController(this, animated) {
+                logger.info(TAG, "Dismissing alert dialog with title: ${alert.title}")
```
Completion of presentViewController does not mean the alert is dismissed. It means that the Alert is now being presented (i.e. the animation has completed)

ChristoferAlexander replied:
Yes indeed (should have read the docs, lucky you knew that). Adjusted to take the event from Action click event for dismissing.

Daeda88 reviewed (changes requested):
Why isnt logger exposed to the BaseAlertPresenter to log some common logs there

Daeda88 commented on alerts/src/iosMain/kotlin/AlertPresenter.kt:
```diff
@@ -119,20 +120,26 @@ actual class AlertPresenter(
                     ) {
-                        logger.info(TAG, "Action ${action.title} was called on dialog with title: ${alert.title}")
+                        if (action.isCancelAction()) {
+                            logger.info(TAG, "Dismissing alert dialog with title: ${alert.title}")
```
You dont know this, a custom action may have a negative or cancel style. Also all actions dismiss

ChristoferAlexander replied:
You are right, I will adjust to log dismissal here. I was trying to figure out a way to add an onDismiss callback but I could not find any. If you know where to point me I think that would be a better solution.

ChristoferAlexander replied:
Changed to that but I see issue with logs not called. For Android dismiss logs are not called when dismissing via action. On iOS appear / dismiss logs are not displayed but click action logs are. Will investigate more.

ChristoferAlexander replied:
I think it looks ok, logs for iOS:

Displaying alert dialog with title: Hello, Kaluga 🐟
10-11 18:28:14.067 💙 INFO AlertDialog - Action Cancel was called on dialog with title: Hello, Kaluga 🐟
10-11 18:28:15.466 💙 INFO AlertDialog - Displaying alert dialog with title: Wait for 3 sec...
10-11 18:28:18.483 💙 INFO AlertDialog - Dismissing alert dialog with title: Wait for 3 sec...
10-11 18:28:20.165 💙 INFO AlertDialog - Displaying alert dialog with title: Wait for 3 sec...
10-11 18:28:21.520 💙 INFO AlertDialog - Action OK was called on dialog with title: Wait for 3 sec...
10-11 18:28:23.432 💙 INFO AlertDialog - Displaying alert dialog with title: Select an option
10-11 18:28:24.990 💙 INFO AlertDialog - Action Option 3 was called on dialog with title: Select an option

And for Android:

Displaying alert dialog with title: Hello, Kaluga 🐟
Action OK was called on dialog with title: Hello, Kaluga 🐟
Displaying alert dialog with title: Hello, Kaluga 🐟
Action Cancel was called on dialog with title: Hello, Kaluga 🐟
Displaying alert dialog with title: Wait for 3 sec...
Dismissing alert dialog with title: Wait for 3 sec...
Displaying alert dialog with title: Wait for 3 sec...
Action OK was called on dialog with title: Wait for 3 sec...
Displaying alert dialog with title: Select an option
Action Option 2 was called on dialog with title: Select an option
Displaying alert dialog with title: Hello, Kaluga 🐟
Action OK was called on dialog with title: Hello, Kaluga 🐟
Displaying alert dialog with title: Hello, Kaluga 🐟
Action Cancel was called on dialog with title: Hello, Kaluga 🐟
Displaying alert dialog with title: Wait for 3 sec...
Dismissing alert dialog with title: Wait for 3 sec...
Displaying alert dialog with title: Wait for 3 sec...
Action OK was called on dialog with title: Wait for 3 sec...
Displaying alert dialog with title: Select an option
Canceling alert dialog with title: Select an option
Displaying alert dialog with title: Select an option
Action Option 3 was called on dialog with title: Select an option

Daeda88 commented on alerts/src/commonMain/kotlin/Alerts.kt:
```diff
@@ -339,13 +345,17 @@ abstract class BaseAlertPresenter(private val alert: Alert) : AlertActions {
+    }
 
-    protected abstract fun showAlert(
+    protected open fun showAlert(
```
I prefer not to have an open fun, but more a structure of
```
protected fun showAlert() {
    logger.info(TAG, "Displaying alert dialog with title: ${alert.title}")
    handleShowAlert()
}

protected abstract fun handleShowAlert()
```

Daeda88 commented on alerts/src/androidLibMain/kotlin/AlertPresenter.kt:
```diff
@@ -123,7 +134,10 @@ actual class AlertPresenter(
                 setItems(titles) { _, which ->
-                    val action = alert.actions[which].apply { handler() }
+                    val action = alert.actions[which].apply {
+                        logger.info(TAG, "Action ${this.title} was called on dialog with title: ${alert.title}")
```
Why cant these just be done in `presentation.afterHandler()`

ChristoferAlexander replied:
Indeed, changed to that.

thoutbeckers commented:
I see comments are resolved, can you rereview @Daeda88 ?

Daeda88 commented:
I've refactored the builder structure a bit, so the Logger is set on the common level instead of the platform level, which I prefer for common APIs (and also makes actually hooking this up to Swift much easier).

Approved, but best @ChristoferAlexander does a final check before merge

ChristoferAlexander commented:
> I've refactored the builder structure a bit, so the Logger is set on the common level instead of the platform level, which I prefer for common APIs (and also makes actually hooking this up to Swift much easier).
> 
> Approved, but best @ChristoferAlexander does a final check before merge

Much appreciated, will run on both platforms and merge. I think I added on the non common part as my original though was that the object is created usually on platform specific level.

ChristoferAlexander commented:
@Daeda88 from my understanding after your changes you will always have to pass the the logger instance when you call 

`builder.buildAlert(this, logger)` which you will have yo do for every single dialog call. Wouldn't it better to pass the logger when we create the instance of the builder in DI?
